sdk/trace/provider.go

```diff
@@ -330,6 +335,32 @@ func WithSpanLimits(sl SpanLimits) TracerProviderOption {
 	})
 }
 
+// WithClock returns a TracerProviderOption that will configure the
+// TracerProvider's clock. The configured clock is used by Tracers
+// to generate span start/end time.
+// If this option is not used, the TracerProvider will provide the default
+// clock which just calls the time package. For the custom Clock `Now`
+// must be implemented (which is used in start time generation). Also user
+// can implement `Since(time.Time) time.Duration` to control how end time is
+// generated (fallback to call `clk.Now().Sub(t)` if not provided).
+func WithClock(clk Clock) TracerProviderOption {
+	rclk := &clockWrapper{
+		nowFunc: clk.Now,
+	}
+	if x, ok := clk.(interface {
+		Since(t time.Time) time.Duration
+	}); ok {
```

Comment:
It seems like we could better document this in code by having this exported as an exported type. Possibly adding another "`Sincer`" (or some better name) interface type that we reference in the docs like "if clk implements Sincer the span end time will be generated with it, otherwise ...". Having this buried in implementation and the option documentation seems like a good way for users to miss this.

Reply:
Although this `Since` method has already been mentioned in the comment, I agree that exposing one more interface would make it more explicit and clear. However there is a little bit of differences from what pellared initially suggested. @pellared What do you think? If that's OK I have no trouble making this change.

Reply:
I agree with @MrAlias.
I see it is even more popular in Go standard library. Examples:
- https://golang.org/pkg/encoding/json/#Unmarshaler
- https://golang.org/pkg/encoding/
- https://golang.org/pkg/encoding/gob/#GobDecoder